### PR TITLE
8291956: Simplify the loop condition in sun.net.httpserver.Request constructor

### PR DESCRIPTION
**Pull request by thyecust**
The condition at line 57 (after while) will evaluate to false
if startLine == null, so the previous if-condition is covered.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291956](https://bugs.openjdk.org/browse/JDK-8291956): Simplify the loop condition in sun.net.httpserver.Request constructor

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Chris Hegarty](https://openjdk.org/census#chegar) (@ChrisHegarty - **Reviewer**) ⚠️ Review applies to [55f9a39b](https://git.openjdk.org/jdk/pull/9571/files/55f9a39b6674ac65abea3a276d9c3d1435dfcfe6)

### Contributors
 * Tian Haoyu `<tienhoayu@gmail.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9571/head:pull/9571` \
`$ git checkout pull/9571`

Update a local copy of the PR: \
`$ git checkout pull/9571` \
`$ git pull https://git.openjdk.org/jdk pull/9571/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9571`

View PR using the GUI difftool: \
`$ git pr show -t 9571`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9571.diff">https://git.openjdk.org/jdk/pull/9571.diff</a>

</details>

### CODE REVIEW

**Comment by jaikiran** [comment]
Hello @thyecust, I've created a JBS issue to track this change https://bugs.openjdk.org/browse/JDK-8291956. Please change the title of this PR to `8291956: Simplify the loop condition in sun.net.httpserver.Request constructor` so that the review process gets initiated.

**Comment by thyecust** [comment]
Thanks for review.

I run `make test-test-tier2` and got an error:

```
...
Compiling 8 files for jdk.unsupported.desktop
In file included from /home/tianh/github/jdk/src/hotspot/share/utilities/globalDefinitions_gcc.hpp:35,
                 from /home/tianh/github/jdk/src/hotspot/share/utilities/globalDefinitions.hpp:35,
                 from /home/tianh/github/jdk/src/hotspot/share/memory/allocation.hpp:29,
                 from /home/tianh/github/jdk/src/hotspot/share/classfile/classLoaderData.hpp:28,
                 from /home/tianh/github/jdk/src/hotspot/share/precompiled/precompiled.hpp:34:
In function 'const void* memchr(const void*, int, size_t)',
    inlined from 'int SignatureStream::scan_type(BasicType)' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.cpp:343:32,
    inlined from 'void SignatureStream::next()' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.cpp:373:19,
    inlined from 'void SignatureIterator::do_parameters_on(T*) [with T = Fingerprinter]' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.hpp:635:41,
    inlined from 'void SignatureIterator::do_parameters_on(T*) [with T = Fingerprinter]' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.hpp:629:6,
    inlined from 'void Fingerprinter::compute_fingerprint_and_return_type(bool)' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.cpp:169:19:
/usr/include/string.h:102:27: error: 'void* __builtin_memchr(const void*, int, long unsigned int)' specified bound [18446744073709486082, 0] exceeds maximum object size 9223372036854775807 [-Werror=stringop-overread]
  102 |   return __builtin_memchr (__s, __c, __n);
      |          ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[3]: *** [lib/CompileJvm.gmk:149: /home/tianh/github/jdk/build/linux-x86_64-server-release/hotspot/variant-server/libjvm/objs/signature.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [make/Main.gmk:252: hotspot-server-libs] Error 2

ERROR: Build failed for target 'test-tier2' in configuration 'linux-x86_64-server-release' (exit code 2) 

=== Output from failing command(s) repeated here ===
* For target hotspot_variant-server_libjvm_objs_signature.o:
In file included from /home/tianh/github/jdk/src/hotspot/share/utilities/globalDefinitions_gcc.hpp:35,
                 from /home/tianh/github/jdk/src/hotspot/share/utilities/globalDefinitions.hpp:35,
                 from /home/tianh/github/jdk/src/hotspot/share/memory/allocation.hpp:29,
                 from /home/tianh/github/jdk/src/hotspot/share/classfile/classLoaderData.hpp:28,
                 from /home/tianh/github/jdk/src/hotspot/share/precompiled/precompiled.hpp:34:
In function 'const void* memchr(const void*, int, size_t)',
    inlined from 'int SignatureStream::scan_type(BasicType)' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.cpp:343:32,
    inlined from 'void SignatureStream::next()' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.cpp:373:19,
    inlined from 'void SignatureIterator::do_parameters_on(T*) [with T = Fingerprinter]' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.hpp:635:41,
    inlined from 'void SignatureIterator::do_parameters_on(T*) [with T = Fingerprinter]' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.hpp:629:6,
    inlined from 'void Fingerprinter::compute_fingerprint_and_return_type(bool)' at /home/tianh/github/jdk/src/hotspot/share/runtime/signature.cpp:169:19:
/usr/include/string.h:102:27: error: 'void* __builtin_memchr(const void*, int, long unsigned int)' specified bound [18446744073709486082, 0] exceeds maximum object size 9223372036854775807 [-Werror=stringop-overread]
  102 |   return __builtin_memchr (__s, __c, __n);
      |          ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

* All command lines available in /home/tianh/github/jdk/build/linux-x86_64-server-release/make-support/failure-logs.
=== End of repeated output ===

No indication of failed target found.
HELP: Try searching the build log for '] Error'.
HELP: Run 'make doctor' to diagnose build problems.
```

Currently there is a patch trying to solve this error: #9711 .

Should I try to run test-tier2 in other ways?

**Comment by dfuch** [comment]
Thanks for trying :-). I have launched a tier2 job for you. I'll approve the PR if results are successful.

**Review by dfuch — APPROVED** [review]
Test results came back green. Good to go!

File: src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java [hunk 1]
```diff
@@ -50,11 +50,8 @@ class Request {
         os = rawout;
         do {
             startLine = readLine();
-            if (startLine == null) {
-                return;
-            }
             /* skip blank lines */
-        } while (startLine == null ? false : startLine.equals (""));
+        } while (startLine != null && startLine.equals (""));
```
Comment (RogerRiggs) [inline src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java:54]: Drive-by comment;  Swapping the arguments for String.equals will handle the case of startLine == null.
`} while ("".equals(startLine));`

Comment (dfuch) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java:54]: Right - which reminds me `String.equals("")` is an anti-pattern that should probably be replaced by `String.isEmpty()`

Comment (dfuch) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java:54]: One or the other :-)

Comment (ChrisHegarty) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java:54]: LGTM.

We need a Strings utility class! ;-) Strings.isEmptyOrNull !

**Review by ChrisHegarty — APPROVED** [review]

File: src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java (hunk 1 above)
Comment (RogerRiggs) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java:54]: `java.util.Objects.equals("", startLine)` would work as well. It covers any type.

Comment (thyecust) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/Request.java]: @RogerRiggs Thanks for your advice :smiley:

**Review by dfuch — APPROVED** [review]
